add CGNAT range to Caddy and document support for additional trusted proxies in rp docs

asavageiv opened the pull request:
When a reverse proxy is on a remote host that proxies from a public IP to a VPN IP (e.g. a Tailscale node) then the VPN IP needs to be a trusted proxy to correctly forward the client IP. This permits that.

asavageiv commented:
Thanks for the reviews!

I also see that this conflicts with `ADDITIONAL_TRUSTED_PROXY` being set by `CADDY_IP_ADDRESS`, so this won't work as-is.

My initial PR description is incomplete and the name `ADDITIONAL_TRUSTED_PROXY` is misleading. Using `ADDITIONAL_TRUSTED_PROXY` because it also adds the IP to the `wopi_allowlist` in `Containers/nextcloud/entrypoint.sh`. It may be that adding it to the WOPI allow list is not necessary if #6676 gets merged. I'm not sure.

I wanted an environment variable to keep a clean and clear interface for supporting reverse proxies. That way if internal things change, like the need for it to be on the allow list, users won't have to make a change and us developers won't have to try and communicate that change. It parallels the way `APACHE_ADDITIONAL_NETWORK` is used for RPs on the same host.

But, I could be even more explicit about the intent with the environment variable name with something like `REVERSE_PROXY_IP`. That would give us a clean interface and maybe allow us to clean up some of the allow list and trusted proxy code internally in the future.

I'm fine with either of these options:
1. (preferred) Use a `REVERSE_PROXY_IP` environment variable in `compose.yaml`
2. Update the RP docs to say "if you're running a RP on a remote host, adjust `trusted_proxies` and the `WOPI allowlist` like this: ..." 

Let me know what you prefer @szaimen. There may be other reasons to prefer one or the other that I'm missing.

asavageiv commented:
I forgot to mention, the external proxy IP also needs to be in the apache container Caddyfile. If it is only set in Nextcloud then the internal Caddy will have its IP detected as `127.0.0.1`, so I'm not sure how to pass it there if it's not an ENV variable.

szaimen commented:
> I forgot to mention, the external proxy IP also needs to be in the apache container Caddyfile. If it is only set in Nextcloud then the internal Caddy will have its IP detected as `127.0.0.1`, so I'm not sure how to pass it there if it's not an ENV variable.

I am honestly not sure about this. This is only needed in your custom setup because you use a public ip-address for the connection between RP and AIO instance IIRC...

szaimen commented:
> Thanks for the reviews!
> 
> I also see that this conflicts with `ADDITIONAL_TRUSTED_PROXY` being set by `CADDY_IP_ADDRESS`, so this won't work as-is.
> 
> My initial PR description is incomplete and the name `ADDITIONAL_TRUSTED_PROXY` is misleading. Using `ADDITIONAL_TRUSTED_PROXY` because it also adds the IP to the `wopi_allowlist` in `Containers/nextcloud/entrypoint.sh`. It may be that adding it to the WOPI allow list is not necessary if #6676 gets merged. I'm not sure.
> 
> I wanted an environment variable to keep a clean and clear interface for supporting reverse proxies. That way if internal things change, like the need for it to be on the allow list, users won't have to make a change and us developers won't have to try and communicate that change. It parallels the way `APACHE_ADDITIONAL_NETWORK` is used for RPs on the same host.
> 
> But, I could be even more explicit about the intent with the environment variable name with something like `REVERSE_PROXY_IP`. That would give us a clean interface and maybe allow us to clean up some of the allow list and trusted proxy code internally in the future.
> 
> I'm fine with either of these options:
> 
> 1. (preferred) Use a `REVERSE_PROXY_IP` environment variable in `compose.yaml`
> 2. Update the RP docs to say "if you're running a RP on a remote host, adjust `trusted_proxies` and the `WOPI allowlist` like this: ..."
> 
> Let me know what you prefer @szaimen. There may be other reasons to prefer one or the other that I'm missing.

I need to think about this...

asavageiv commented:
I drew my set up below for clarification. I'm surprised this is not a common configuration.

I updated the PR to allow `ADDITIONAL_TRUSTED_PROXY` xor the Caddy community container and removed `CADDY_IP_ADDRESS` placeholder in favor of returning the `CADDY_IP_ADDRESS` value as the value of `ADDITIONAL_TRUSTED_PROXY` if the community container is enabled. This should keep things working for the community container and solve my issue.

![asavageiv nextcloud](https://github.com/user-attachments/assets/3eff97fb-96cb-480b-bd07-e8f27f1d6e6e)

szaimen commented:
> > I'm fine with either of these options:
> > 
> > 1. (preferred) Use a `REVERSE_PROXY_IP` environment variable in `compose.yaml`
> > 2. Update the RP docs to say "if you're running a RP on a remote host, adjust `trusted_proxies` and the `WOPI allowlist` like this: ..."
> > 
> > Let me know what you prefer @szaimen. There may be other reasons to prefer one or the other that I'm missing.
> 
> I need to think about this...

Ok, thought about this now and I would prefer option 2 :)

szaimen commented on Containers/apache/start.sh:
```diff
@@ -46,7 +46,11 @@ echo "$CADDYFILE" > /tmp/Caddyfile
 
 # Change the trusted_proxies in case of reverse proxies
 if [ "$APACHE_PORT" != '443' ]; then
+  if [ -z "$ADDITIONAL_TRUSTED_PROXY" ]; then
     CADDYFILE="$(sed 's|# trusted_proxies placeholder|trusted_proxies static private_ranges|' /tmp/Caddyfile)"
+  else
+    CADDYFILE="$(sed "s|# trusted_proxies placeholder|trusted_proxies static private_ranges $ADDITIONAL_TRUSTED_PROXY|" /tmp/Caddyfile)"
+  fi
```
So basically add the range here:
```suggestion
CADDYFILE="$(sed 's|# trusted_proxies placeholder|trusted_proxies static private_ranges 100.64.0.0/10|' /tmp/Caddyfile)"
```

asavageiv commented:
> I dont have a starlink connection for testing but it should security wise not be possible to access other ip-adresses inside the ip range anyway, no?

It's up to the ISP. It is common for other IPs within the range to be reachable. Whether or not they have properly configured firewalls is another thing, but for AIO users if they have `APACHE_PORT: 11000` and `APACHE_IP_BINDING: 0.0.0.0` and their ISP doesn't isolate the addresses then their server would be accessible. Someone would still have to set up a reverse proxy with the correct domain name to connect to it, so maybe this is an acceptable risk.

szaimen commented:
> so maybe this is an acceptable risk.

Yes, I think so 👍

szaimen commented on reverse-proxy.md:
```diff
@@ -990,7 +990,19 @@ After starting AIO, you should be able to access the AIO Interface via `https://
+
+<small>*: For example, the reverse proxy has a public globally routable IP and connects to your AIO instance via Tailscale with an IP in the `100.64.0.0/10` range.</small>
+
+## 6. Optional: get a valid certificate for the AIO interface
```
we have the problem now that adjusting the heading will break old links from external to this section. This is probably not a big issue with this specific section but it is for `## 6. How to debug things?` which should be adjusted to `## 7. How to debug things?`. Do you have an idea how we can solve this?

szaimen replied:
The link then also needs to be adjusted in these places: https://github.com/search?q=repo%3Anextcloud%2Fall-in-one%20how-to-debug-things&type=code

asavageiv replied:
This is a great point and something I hadn't considered. What we do depends a bit on the requirements for external linking. Do we want the whole web to be able to link to a section in this document? Just the "how to debug" section? Or just within the repository or Nextcloud organization?

We want to be able to reword and reorganize the document, but we also want somewhat stable links.

There are a few options:
1. Append a new section to the end or make it a subsection of an existing section. Easy right now but might make the doc harder to read over time. So I don't like this one. I want minimal barriers to making the documentation clearer.
1. Insert it, but add anchors to preserve backwards compatibility. E.g. 
    ```
    ## 7. How to debug things?
    <a id="6-how-to-debug-things"></a>
    ```
    Over time could lead to cruft build-up in the docs unless we prefer manually creating stable links over the auto-generated ones in the future. That is, in the future put `<a id="how-to-debug"` on things externally linked so there's no reference to the document structure in the link, just the content. It has the advantage of maintaining links on the WWW. We could also have both links and remove the old ones in the future.
1. Insert it and update all the links we know about. This breaks all links outside of the repo and leaves us always having to remember to check and search for any link changes in any documentation we update.

I like the idea of an explicit anchor for anything we want linked from outside the document. `<a id="how-to-debug">` makes it obvious "this thing is a stable link and shouldn't be changed" but lets us reorganize the document in the future. We probably want at least temporary backwards compatibility with error message links from older AIO versions, so I put in an additional `<a id="6-how-to-debug-things"></a><!-- for backwards compatibility -->` which we can remove later.

There might be other ways I'm not thinking of, but I think you get the idea.

asavageiv commented on reverse-proxy.md:
```diff
@@ -990,7 +991,28 @@ After starting AIO, you should be able to access the AIO Interface via `https://
+### Collabora WOPI allow list
+If your reverse proxy connects to Nextcloud with an IP address that is different from the one for your domain<sup>*</sup> and you are using the Collabora server then you must also add the IP to the WOPI request allow list via `Administration Settings > Administration > Office > Allow list for WOPI requests`.
+
+<small>*: For example, the reverse proxy has a public globally routable IP and connects to your AIO instance via Tailscale with an IP in the `100.64.0.0/10` range.</small>
```
Agreed, I think it's a good idea. Added a proposed small change.

szaimen commented on reverse-proxy.md:
````diff
@@ -990,7 +991,28 @@ After starting AIO, you should be able to access the AIO Interface via `https://
+```
+sudo docker exec --user www-data -it nextcloud-aio-nextcloud php occ config:system:set trusted_proxies 2 --value=ip.address.of.proxy
+```
+
````
```suggestion

Advice: In case you run AIO behind Cloudflare Tunnel, you must add all [Cloudflare IP-Ranges](https://www.cloudflare.com/ips/) to the WOPI allowlist.
